src/main/java/org/apache/sling/xss/impl/CustomPolicy.java

```diff
@@ -0,0 +1,265 @@
+            }
+
+            if (CssValidator.STYLE_ATTRIBUTE_NAME.equals(attribute.getName())) {
+                // we match style tags separately
+                policyBuilder.allowAttributes(attribute.getName()).matching(cssValidator.newCssAttributePolicy())
+                        .globally();
+            } else {
+                List<String> allowedValuesFromAttribute = attribute.getLiterals();
+                if (allowedValuesFromAttribute != null && allowedValuesFromAttribute.size() > 0) {
+                    for (String allowedValue : allowedValuesFromAttribute) {
+                        policyBuilder.allowAttributes(attribute.getName()).matching(true, allowedValue).globally();
+                    }
+
+                }
+                List<Pattern> regexsFromAttribute = attribute.getPatternList();
+                if (regexsFromAttribute != null && regexsFromAttribute.size() > 0) {
+                    policyBuilder.allowAttributes(attribute.getName()).matching(matchesToPatterns(regexsFromAttribute))
+                            .globally();
+                } else {
+                    policyBuilder.allowAttributes(attribute.getName()).globally();
+                }
+
+            }
+        }
+
+        // ------------ this is for the allowed emty tags -------------
+        List<String> allowedEmptyTags = policy.getAllowedEmptyTags();
+        for (String allowedEmptyTag : allowedEmptyTags) {
+            policyBuilder.allowWithoutAttributes(allowedEmptyTag);
+        }
+
+        // ------------ this is for the tag rules -------------
+        Map<String, Tag> tagMap = policy.getTagRules();
+        for (Map.Entry<String, Tag> tag : tagMap.entrySet()) {
+
+            String tagAction = tag.getValue().getAction();
+            switch (tagAction) {
+            // Tag.action
+            case "truncate":
+                policyBuilder.allowElements(tag.getValue().getName());
+
+                break;
+            // filter: remove tags, but keep content,
+            case "filter":
+                break;
+            // remove: remove tag and contents
+            case "remove":
+                policyBuilder.disallowElements(tag.getValue().getName());
+                break;
+
+            // validate is also the default
+            // validate: keep content as long as it passes rules,
+            default:
+                policyBuilder.allowElements(tag.getValue().getName());
+                boolean styleSeen = false;
+                // get the allowed Attributes for the tag
+                Map<String, Attribute> allowedAttributes = tag.getValue().getAttributeMap();
+                if (allowedAttributes != null && allowedAttributes.size() > 0) {
+
+                    // if there are allowed Attributes, map over them
+                    for (Attribute attribute : allowedAttributes.values()) {
+                        if (attribute.getOnInvalid().equals("removeTag")) {
```

Comment:
We should extract these string values as constants somewhere and reference them from the code.

Reply:
Resolved: 36af1dd88775dfa5e6e2e76712bbbd9212d751ab